libvast/src/system/pivoter.cpp

```diff
@@ -0,0 +1,159 @@
+ ******************************************************************************/
+
+#include "vast/system/pivoter.hpp"
+
+#include "vast/command.hpp"
+#include "vast/concept/parseable/to.hpp"
+#include "vast/concept/parseable/vast/expression.hpp"
+#include "vast/detail/string.hpp"
+#include "vast/expression.hpp"
+#include "vast/logger.hpp"
+#include "vast/system/exporter.hpp"
+#include "vast/table_slice.hpp"
+
+#include <caf/atom.hpp>
+#include <caf/event_based_actor.hpp>
+
+namespace vast::system {
+
+namespace {
+
+caf::optional<record_field>
+common_field(const pivoter_state& st, const record_type& indicator) {
+  auto f = st.cache.find(indicator);
+  if (f != st.cache.end())
+    return f->second;
+    // TODO: This algorithm can be enabled once we have a live updated
+    //       type registry. (Switch the type of target to record_type.)
+#if 0
+  for (auto& t : target.fields) {
+    for (auto& i : indicator.fields) {
+      if (t.name == i.name) {
+        st.cache.insert({indicator, i});
+        return i;
+      }
+    }
+  }
+#else
+  std::string edge;
+  VAST_TRACE(st.self, VAST_ARG(st.target), VAST_ARG(indicator.name()));
+  if (detail::starts_with(st.target, "suricata") || st.target == "pcap.packet")
+    edge = "community_id";
+  else if (detail::starts_with(st.target, "zeek")) {
+    if (detail::starts_with(indicator.name(), "zeek"))
+      edge = "uid";
+    else
+      edge = "community_id";
+  }
+  for (auto& i : indicator.fields) {
+    if (i.name == edge) {
+      st.cache.insert({indicator, i});
+      return i;
+    }
+  }
+#endif
+  st.cache.insert({indicator, caf::none});
+  VAST_WARNING(st.self, "got slice without shared column:", indicator.name());
+  return caf::none;
+}
+
+} // namespace
+
+pivoter_state::pivoter_state(caf::event_based_actor*) {
+  // nop
+}
+
+caf::behavior pivoter(caf::stateful_actor<pivoter_state>* self, caf::actor node,
+                      std::string target, expression expr) {
+  auto& st = self->state;
+  st.self = self;
+  st.node = node;
+  st.expr = std::move(expr);
+  st.target = std::move(target);
+  auto quit_if_done = [=]() {
+    auto& st = self->state;
+    if (st.initial_completed && st.outstanding_requests == 0)
+      self->quit();
+  };
+  self->set_down_handler([=](const caf::down_msg& msg) {
+    // Only the spawned EXPORTERs are expected to send down messages.
+    auto& st = self->state;
+    st.outstanding_requests--;
+    VAST_DEBUG(self, "received DOWN from", msg.source,
+               "outstanding requests:", st.outstanding_requests);
+    quit_if_done();
+  });
+  return {[=](vast::table_slice_ptr slice) {
+            auto& st = self->state;
+            auto pivot_field = common_field(st, slice->layout());
+            if (!pivot_field)
+              return;
+            VAST_DEBUG(self, "uses", *pivot_field, "to extract", st.target,
+                       "events");
+            auto pivot_field_name = pivot_field->name;
```

Comment:
I don't think this alias adds much; I'd keep `pivot_field->name` in the remaining code.